src/appleseed.studio/mainwindow/rendering/lightpathswidget.cpp

```diff
@@ -0,0 +1,382 @@
+    {
+        const Matrix4d mt(transpose(m));
+        ::glMultMatrixd(const_cast<GLdouble*>(&mt[0]));
+    }
+
+    struct OpenGLRasterizer
+      : public ObjectRasterizer
+    {
+        void begin_object() override
+        {
+            glBegin(GL_TRIANGLES);
+        }
+
+        void end_object() override
+        {
+            glEnd();
+        };
+
+        void rasterize(const Triangle& triangle) override
+        {
+            glNormal3d(triangle.m_n0[0], triangle.m_n0[1], triangle.m_n0[2]);
+            glVertex3d(triangle.m_v0[0], triangle.m_v0[1], triangle.m_v0[2]);
+
+            glNormal3d(triangle.m_n1[0], triangle.m_n1[1], triangle.m_n1[2]);
+            glVertex3d(triangle.m_v1[0], triangle.m_v1[1], triangle.m_v1[2]);
+            
+            glNormal3d(triangle.m_n2[0], triangle.m_n2[1], triangle.m_n2[2]);
+            glVertex3d(triangle.m_v2[0], triangle.m_v2[1], triangle.m_v2[2]);
+        }
+    };
+
+    void draw(const ObjectInstance& object_instance)
+    {
+        Object* object = object_instance.find_object();
+
+        if (object == nullptr)
+            return;
+
+        static const GLfloat Ambient[] = { 0.0f, 0.0f, 0.0f, 1.0f };
+        static const GLfloat Diffuse[] = { 0.8f, 0.8f, 0.8f, 1.0f };
+        static const GLfloat Specular[] = { 0.4f, 0.4f, 0.4f, 1.0f };
+        static const GLfloat Shininess = 20.0f;
+
+        glMaterialfv(GL_FRONT_AND_BACK, GL_AMBIENT, Ambient);
+        glMaterialfv(GL_FRONT_AND_BACK, GL_DIFFUSE, Diffuse);
+        glMaterialfv(GL_FRONT_AND_BACK, GL_SPECULAR, Specular);
+        glMaterialf(GL_FRONT_AND_BACK, GL_SHININESS, Shininess);
+
+        const Transformd& transform = object_instance.get_transform();
+
+        glPushMatrix();
+        glMultMatrixd(transform.get_local_to_parent());
+
+        OpenGLRasterizer rasterizer;
+        object->rasterize(rasterizer);
+
+        glPopMatrix();
+    }
+
+    void draw(
+        const AssemblyInstance& assembly_instance,
+        const float             time)
+    {
+        const Assembly* assembly = assembly_instance.find_assembly();
+
+        if (assembly == nullptr)
+            return;
+
+        const Transformd transform = assembly_instance.transform_sequence().evaluate(time);
+
+        glPushMatrix();
+        glMultMatrixd(transform.get_local_to_parent());
+
+        for (const auto& object_instance : assembly->object_instances())
+            draw(object_instance);
+
+        for (const auto& child_assembly_instance : assembly->assembly_instances())
+            draw(child_assembly_instance, time);
+
+        glPopMatrix();
+    }
+}
+
+void LightPathsWidget::paintGL()
+{
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+    const auto& rc = m_camera.get_rasterization_camera();
+
+    glMatrixMode(GL_PROJECTION);
+    glLoadIdentity();
+    gluPerspective(
+        rad_to_deg(rc.m_hfov) / rc.m_aspect_ratio,
+        rc.m_aspect_ratio,
+        0.01,
+        1000.0);
```

Comment:
Far and near z should be configurable

Reply:
I completely agree but I needed to get this feature out of the door early. Over time we will add ways to configure OpenGL rendering including near and far Z, shading mode, etc.